# Actividad 4 | Métricas de calidad de resultados
---

**MAESTRÍA EN INTELIGENCIA ARTIFICIAL APLICADA**

**Curso: TC4034.10 - Análisis de grandes volúmenes de datos**

Tecnológico de Monterrey

* Dr. Iván Olmos Pineda
* Mtra. Verónica Sandra Guzmán de Valle
* Mtro. Alberto Daniel Salinas Montemayor

**Actividad 4** - 
Métricas de calidad de resultados

**Fecha de entrega** - 
8 de junio del 2025

---

**Presenta**

|  NOMBRE COMPLETO                        |     MATRÍCULA     |
| :-------------------------------------: |:-----------------:|
| Alejandro Díaz Villagómez               |  A01276769        |

# 0) Carga de datos
---

In [1]:
import findspark
from pyspark.sql import SparkSession, functions as F
from itertools import product
from os import path

findspark.init()
findspark.find()

'/Users/alejandrodiazvillagomez/Desktop/Proyecto-Big-Data-PySpark/.venv/lib/python3.12/site-packages/pyspark'

In [2]:
spark = SparkSession.builder.master("local[*]").getOrCreate()
spark.sparkContext.setLogLevel("ERROR")
spark.conf.set("spark.sql.repl.eagerEval.enabled", True)

spark

25/06/01 20:07:43 WARN Utils: Your hostname, Alejandros-MacBook-Pro.local resolves to a loopback address: 127.0.0.1; using 192.168.100.202 instead (on interface en0)
25/06/01 20:07:43 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
25/06/01 20:07:44 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


In [3]:
class FileManager():
    @staticmethod
    def open_csv_file(file_path : str):
        """
        This method opens a csv file with pyspark
        """
        csv_df = spark.read.csv(
            file_path,
            header=True,
            inferSchema=True,
            multiLine=True,
            escape="\"",
            quote="\""
        )

        # csv_df.show(truncate=20)

        return csv_df

In [4]:
import kagglehub

# Download latest version
FILE_PATH = kagglehub.dataset_download("machharavikiran/amazon-reviews")

print("Path to dataset files:", FILE_PATH)

df_reviews = FileManager.open_csv_file(FILE_PATH)
df_reviews.limit(10).toPandas()

/Users/alejandrodiazvillagomez/Desktop/Proyecto-Big-Data-PySpark/.venv/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Path to dataset files: /Users/alejandrodiazvillagomez/.cache/kagglehub/datasets/machharavikiran/amazon-reviews/versions/1


marketplace  customer_id       review_id  product_id  product_parent  \
0          US     22873041  R3ARRMDEGED8RD  B00KJWQIIC       335625766   
1          US     30088427   RQ28TSA020Y6J  B013ALA9LA       671157305   
2          US     20329786   RUXJRZCT6953M  B00PML2GQ8       982036237   
3          US     14215710   R7EO0UO6BPB71  B001NS0OZ4       576587596   
4          US     38264512  R39NJY2YJ1JFSV  B00AQMTND2       964759214   
5          US     30548466  R31SR7REWNX7CF  B00KX4TORI       170101802   
6          US       589298   RVBP8I1R0CTZ8  B00P17WEMY       206124740   
7          US     49329488  R1QF6RS1PDLU18  B00TR05L9Y       778403103   
8          US     50728290  R23AICGEDAJQL1  B0098Y77OG       177098042   
9          US     37802374  R2EY3N4K9W19UP  B00IFYEYXC       602496520   

                                       product_title product_category  \
0  Plemo 14-Inch Laptop Sleeve Case Waterproof Fa...               PC   
1         TP-Link OnHub AC1900 Wireless Wi-Fi Router               PC   
2  AmazonBasics USB 3.0 A Male to A Male Cable - ...               PC   
3  Transcend P8 15-in-1 USB 2.0 Flash Memory Card...               PC   
4  Aleratec SATA Data Cable 2.0 20in Serial ATA S...               PC   
5  Kingston Digital MobileLite G4 USB 3.0 Multi-F...               PC   
6  White 9 Inch Unlocked Dual Sim Card Phone Tabl...               PC   
7  Lenovo TAB2 A10 - 10.1" 2-in-1 Tablet (1.5Ghz,...               PC   
8                                               Acer               PC   
9  AzureWave Broadcom BCM94352HMB 802.11/ac/867Mb...               PC   

   star_rating  helpful_votes  total_votes vine verified_purchase  \
0            5              0            0    N                 Y   
1            5             24           31    N                 N   
2            1              2            2    N                 N   
3            1              0            0    N                 Y   
4            5              0            0    N                 Y   
5            5              0            0    N                 Y   
6            3              1            2    N                 Y   
7            4              1            1    N                 Y   
8            1              0            0    N                 Y   
9            5              3            4    N                 Y   

                                     review_headline  \
0                               Pleasantly surprised   
1  OnHub is a pretty no nonsense type router that...   
2     None of them worked.  No functionality at all.   
3                               just keep searching.   
4                                         Five Stars   
5               Good quality, works well and compact   
6        in fact this is third China good. Demn s***   
7                                               Good   
8                           You get what you pay for   
9                        Great for Windows 7 Laptop!   

                                         review_body review_date  sentiment  
0  I was very surprised at the high quality of th...  2015-08-31          1  
1  I am a Google employee and had to chance to us...  2015-08-31          1  
2  Bought cables in 3ft, 6ft and 9ft.  NONE of th...  2015-08-31          0  
3                               nope, cheap and slow  2015-08-31          0  
4           Excellent! Great value and does the job.  2015-08-31          1  
5           Good quality,works well and compact size  2015-08-31          1  
6  This demn tablet is just a Real Chinese produc...  2015-08-31          0  
7  I am not sure I don't know if it is the tablet...  2015-08-31          1  
8  After exactly 45 days, the screen went dark. P...  2015-08-31          0  
9  Replaced my Intel Centrino 2230 with the BCM94...  2015-08-31          1

# 1) Construcción de la muestra M
---

* **Instrucciones**: Construir una muestra `𝑀` que sea representativa de la población `𝑃` (a partir del dataset que recolectaste desde el inicio del curso). Tomando como base el conocimiento adquirido en la *Actividad 3 del Módulo 4*, generarás particiones `𝑀𝑖` de `𝑀`, donde cada `𝑀𝑖` cumple con los criterios definidos por las variables de caracterización que identificaste previamente (`𝑀` será igual a la unión de todos los `𝑀𝑖`). Para esta actividad, y a diferencia del paso previo, se deberá de tener especial cuidado para determinar el número de instancias que deberá contender cada partición `𝑀𝑖` a generar, de tal forma que no se inyecte ningún tipo de sesgo que pueda alterar la calidad de los resultados.

## 1.1) Analizar la Distribución de las Variables Categóricas

In [ ]:
categorical_cols = ["verified_purchase", "vine"]

print("--- Análisis de la Distribución de la Población (df_reviews) ---")

# Contar las ocurrencias de cada combinación de las variables categóricas
population_distribution = df_reviews.groupBy(categorical_cols).count()
population_distribution = population_distribution.withColumn(
    "percentage", (F.col("count") / df_reviews.count()) * 100
).orderBy(categorical_cols)

print("\nDistribución de las combinaciones en la población:")
population_distribution.toPandas()


--- Análisis de la Distribución de la Población (df_reviews) ---


Distribución de las combinaciones en la población:


verified_purchase vine    count  percentage
0                 N    N   823420   11.922281
1                 N    Y    36069    0.522242
2                 Y    N  6046916   87.553174
3                 Y    Y      159    0.002302

En esta primera etapa del proceso, se realiza un análisis exploratorio de la distribución de la población original (`df_reviews`) con base en dos variables categóricas seleccionadas: `verified_purchase` y `vine`. A diferencia del entregable previo, en esta ocasión se decidió excluir la variable `star_rating` del conjunto de variables de caracterización. Esta decisión responde a un hallazgo clave derivado del análisis supervisado anterior, donde se identificó que `star_rating` guarda una correlación directa y significativa con la variable objetivo `sentiment`, lo cual podría introducir un sesgo indeseado en la construcción de la muestra representativa. Mantener únicamente variables neutras y no directamente relacionadas con la etiqueta de salida garantiza una segmentación más imparcial y metodológicamente sólida.

El análisis de distribución revela que la gran mayoría de las observaciones corresponde a reseñas verificadas (`verified_purchase = Y`) y no asociadas al programa Vine (`vine = N`), representando aproximadamente el 87.55% del total. Otras combinaciones, como reseñas no verificadas (`verified_purchase = N`) o aquellas marcadas como parte del programa Vine (`vine = Y`), conforman proporciones mucho menores. Este paso es fundamental ya que permite identificar las proporciones reales de cada estrato dentro de la población y constituye la base para una posterior selección muestral proporcional, asegurando que la muestra `𝑀` sea verdaderamente representativa de la población `𝑃`.

## 1.2) Determinar el Tamaño de la Muestra (M)

In [ ]:
# Vamos a definir un tamaño de muestra. Para este ejemplo, usemos el 10% del total de registros.
# Puedes ajustar este valor según tus necesidades y recursos.
sample_fraction = 0.1
total_population_count = df_reviews.count()
sample_size = int(total_population_count * sample_fraction)

print(f"Total de registros en la población: {total_population_count}")
print(f"Tamaño de la muestra deseado (aprox. {sample_fraction*100}%): {sample_size}")


Total de registros en la población: 6906564
Tamaño de la muestra deseado (aprox. 10.0%): 690656


En esta segunda etapa se define el tamaño de la muestra `𝑀` que será extraída de la población `𝑃`, con el propósito de construir un subconjunto representativo que preserve la estructura y diversidad de los datos originales. Para este fin, se estableció un valor de muestreo equivalente al **10% del total de registros disponibles**. Esta proporción fue seleccionada con base en criterios metodológicos y computacionales, considerando tanto la escala del conjunto de datos como los objetivos del análisis posterior.

Desde el punto de vista estadístico, una fracción del 10% resulta suficiente para capturar patrones relevantes y garantizar una adecuada representación de las combinaciones categóricas identificadas previamente, incluso en escenarios donde algunos estratos presentan baja frecuencia relativa. Además, esta decisión permite mantener un equilibrio adecuado entre la fidelidad de la muestra y la eficiencia computacional, particularmente importante en contextos de Big Data donde el volumen de información puede impactar significativamente el rendimiento de las operaciones.

Al delimitar explícitamente el tamaño de la muestra, se establece un marco cuantitativo claro que orienta la distribución proporcional de las instancias en cada segmento, evitando sesgos y distorsiones en las etapas siguientes del análisis.

## 1.3) Calcular el Número de Instancias por Partición (Mi)

In [9]:
# Calcular el número de instancias para cada combinación en la muestra
print("--- Distribución Deseada de la Muestra y Conteo por Partición ---")

sample_distribution = population_distribution.withColumn(
    "sample_count", F.round(F.col("percentage") * sample_size / 100)
)
sample_distribution.toPandas() # Muestra el conteo deseado para cada partición en la muestra

# Convertir a un diccionario para facilitar la selección de la muestra estratificada
sample_counts_dict = {
    tuple(row[col] for col in categorical_cols): row["sample_count"]
    for row in sample_distribution.collect()
}
sample_counts_dict

--- Distribución Deseada de la Muestra y Conteo por Partición ---


{('N', 'N'): 82342.0,
 ('N', 'Y'): 3607.0,
 ('Y', 'N'): 604691.0,
 ('Y', 'Y'): 16.0}

Se procede a determinar la distribución deseada de la muestra representativa `𝑀`, en función de las proporciones previamente observadas en la población `𝑃` respecto a las variables categóricas seleccionadas: `verified_purchase` y `vine`. Para ello, se calcula la cantidad de instancias que cada combinación posible de estas variables deberá contener dentro de la muestra, con base en su porcentaje de representación original y el tamaño total previamente definido para `𝑀` (10% del total poblacional).

Este procedimiento tiene como objetivo preservar la estructura interna de la población, garantizando que la muestra conserve de manera proporcional la diversidad y distribución de los distintos perfiles existentes en los datos. En otras palabras, se busca construir una muestra estratificada que refleje fielmente la composición original de los datos en términos de las variables de caracterización elegidas.

Los resultados obtenidos muestran que la combinación más frecuente es `('Y', 'N')`, correspondiente a registros de usuarios que realizaron una compra verificada y que no pertenecen al programa Vine, con una representación esperada de aproximadamente 604,691 registros dentro de la muestra. Le siguen `('N', 'N')` con 82,342 registros, `('N', 'Y')` con 3,607, y `('Y', 'Y')` con solo 16 registros. Esta distribución, aunque desigual, es consistente con la frecuencia real de ocurrencia en la población y será la base para extraer de forma estratificada los registros que compondrán la muestra final `𝑀`.

Este paso es esencial para asegurar que las particiones `𝑀𝑖`, construidas posteriormente a partir de `𝑀`, sean equilibradas, realistas y libres de sesgos de representación, lo que a su vez fortalece la validez estadística de cualquier análisis futuro derivado del conjunto muestral.

# 1.4) Generar la Muestra (M) de forma estratificada

In [ ]:
# Para asegurar la representatividad, realizaremos un muestreo estratificado.
# Esto implica muestrear cada "estrato" (combinación de categorías) por separado para asegurar las proporciones deseadas.
print("--- Generando la Muestra (M) de forma Estratificada ---")

sample_df = spark.createDataFrame([], df_reviews.schema) # DataFrame vacío para la muestra

for combo_values, count_needed in sample_counts_dict.items():
    if count_needed > 0:
        filter_condition = None
        for i, col_name in enumerate(categorical_cols):
            if filter_condition is None:
                filter_condition = (F.col(col_name) == combo_values[i])
            else:
                filter_condition = filter_condition & (F.col(col_name) == combo_values[i])

        # Seleccionar las filas para el estrato actual
        stratum_df = df_reviews.filter(filter_condition)

        # Si el estrato tiene menos registros de los que necesitamos, tomamos todos.
        # Si tiene más, tomamos una muestra aleatoria del tamaño deseado.
        current_stratum_count = stratum_df.count()
        if current_stratum_count > 0:
            if count_needed < current_stratum_count:
                # Tomamos una muestra aleatoria del tamaño deseado
                fraction = count_needed / current_stratum_count
                sampled_stratum = stratum_df.sample(withReplacement=False, fraction=fraction, seed=42)
            else:
                # Si no hay suficientes registros en el estrato, tomamos todos los disponibles
                sampled_stratum = stratum_df

            # Unir al DataFrame de la muestra principal
            sample_df = sample_df.unionAll(sampled_stratum)

print(f"Número de registros en la muestra final (M): {sample_df.count()}")
print("Muestra (M) generada. Mostrando las primeras 5 filas:")
sample_df.limit(5).toPandas()

--- Generando la Muestra (M) de forma Estratificada ---


Número de registros en la muestra final (M): 691829
Muestra (M) generada. Mostrando las primeras 5 filas:


marketplace  customer_id       review_id  product_id  product_parent  \
0          US       112405   RKP98DQX0VTFZ  B00T91YABQ       541159143   
1          US     46558847  R3MSK0V0TDL01L  B00TKFDFW6       701421383   
2          US     19852854   RM6KESMRIY1UQ  B00746W9F2        41875278   
3          US     13443121  R2A948UZQGFZHO  B00KQE99C0       348236562   
4          US       107911   RS1QTAUB8FJI0  B00JQZJ75E       937857628   

                                       product_title product_category  \
0  Polare Thick Full Grain Leather Shoulder Brief...               PC   
1  CybertronPC SOKOM-I Gaming Desktop - AMD FX-63...               PC   
2                                    Apple iPad mini               PC   
3  Standing Protective Case for Fire HD 7 (4th Ge...               PC   
4        iXCC Silicon Case Cover for Apple iPad Mini               PC   

   star_rating  helpful_votes  total_votes vine verified_purchase  \
0            5              1            3    N                 N   
1            5              1            1    N                 N   
2            5              0            0    N                 N   
3            5              0            0    N                 N   
4            5              0            0    N                 N   

          review_headline                                        review_body  \
0            Awesome Bag!  I bought this bag looking for a briefcase that...   
1  It's a Monster Machine  I bought a slightly modified version of this C...   
2              Five Stars                                   very good price!   
3              Five Stars  Love it, it protects the screen and I love the...   
4                 5 stars  Nice case for the money. It is comfortable to ...   

  review_date  sentiment  
0  2015-08-31          1  
1  2015-08-31          1  
2  2015-08-31          1  
3  2015-08-31          1  
4  2015-08-31          1

En este paso se lleva a cabo la generación efectiva de la muestra `𝑀` mediante un proceso de **muestreo estratificado**. A diferencia de un muestreo aleatorio simple, el enfoque estratificado garantiza que cada subconjunto de interés (en este caso, definido por las combinaciones de las variables categóricas `verified_purchase` y `vine`) esté adecuadamente representado en la muestra final, en proporciones equivalentes a las observadas en la población `𝑃`.

Para ello, se recorre cada combinación posible de valores previamente identificada y se extrae un subconjunto de registros desde el conjunto poblacional, aplicando una lógica condicional que asegura tanto la proporcionalidad deseada como la eficiencia computacional. En particular, cuando un estrato contiene más registros que los requeridos, se realiza un muestreo aleatorio sin reemplazo proporcional a su tamaño; en contraste, si el estrato es más pequeño que la cantidad esperada, se conservan todos los registros disponibles. Este tratamiento dual permite construir una muestra robusta, incluso en presencia de categorías minoritarias o desbalanceadas.

El resultado final fue la obtención de una muestra `𝑀` con 691,829 registros, cifra levemente superior al valor estimado originalmente (690,656) debido a efectos de redondeo y disponibilidad real de datos en algunos estratos. Esta ligera discrepancia es aceptable dentro de un marco de análisis exploratorio y no compromete la representatividad del conjunto, ya que la preservación de las proporciones entre categorías sigue siendo el criterio rector de la construcción muestral.

Como se puede apreciar, este paso sienta las bases para asegurar que cualquier inferencia posterior, basada en la muestra `𝑀` o en sus particiones `𝑀𝑖`, no esté sesgada por sobre o subrepresentación de ciertos grupos, manteniendo la integridad metodológica del estudio.

# 1.5) Crear las Particiones (Mi)

In [12]:
# Una vez que tenemos la muestra (sample_df), las particiones Mi son simplemente subconjuntos de esta 
# muestra basados en las combinaciones de las variables categóricas.
# No es necesario almacenar cada Mi por separado si se va a trabajar con ellos de forma iterativa o agrupada. 
# Podemos simplemente agrupar por las variables categóricas cuando sea necesario.

# Si se necesita acceder a una partición específica (Mi), se puede filtrar así:
# Ejemplo: Partición para verified_purchase='Y', vine='N'
mi_example = sample_df.filter((F.col("verified_purchase") == "Y") &
                              (F.col("vine") == "N"))
print(f"\nEjemplo de partición M_i (verified_purchase='Y', vine='N'): {mi_example.count()} registros")
mi_example.limit(5).toPandas()


Ejemplo de partición M_i (verified_purchase='Y', vine='N'): 605672 registros


marketplace  customer_id       review_id  product_id  product_parent  \
0          US     37802374  R2EY3N4K9W19UP  B00IFYEYXC       602496520   
1          US     29333557  R3G4RT3EQ9RSY7  B00MA40W9I       535866197   
2          US     16160990   RW9VR7UKL3E69  B00HPQ0H0K       176437370   
3          US     48701722   R7ER63LS7Q1O5  B00RXEWOAA       540891596   
4          US     19280847    RAYGOWHX4H47  B001J226JQ       210607495   

                                       product_title product_category  \
0  AzureWave Broadcom BCM94352HMB 802.11/ac/867Mb...               PC   
1  Egoway® New Laptop Battery for Apple A1309 A12...               PC   
2  ARCTIC S111 USB-Powered Portable Stereo Speake...               PC   
3  3.5" USB External Floppy Disk Drive Portable 1...               PC   
4           Generic USB to 9-pin Serial Port Adapter               PC   

   star_rating  helpful_votes  total_votes vine verified_purchase  \
0            5              3            4    N                 Y   
1            1              0            0    N                 Y   
2            2              0            0    N                 Y   
3            5              0            0    N                 Y   
4            5              0            0    N                 Y   

                      review_headline  \
0         Great for Windows 7 Laptop!   
1  Totally wasted $60 on this battery   
2                      tinny sounding   
3                          Five Stars   
4                          Five Stars   

                                         review_body review_date  sentiment  
0  Replaced my Intel Centrino 2230 with the BCM94...  2015-08-31          1  
1  Totally wasted $60 on this battery.  Didn't wo...  2015-08-31          0  
2                                   tinny weak sound  2015-08-31          0  
3                                            Love it  2015-08-31          1  
4                                            perfect  2015-08-31          1

In [ ]:
print("--- Creando las Particiones (Mi) ---")

# Verificamos que las proporciones se mantengan.
sample_actual_distribution = sample_df.groupBy(categorical_cols).count()
sample_actual_distribution = sample_actual_distribution.withColumn(
    "percentage", (F.col("count") / sample_df.count()) * 100
).orderBy(categorical_cols)

print("\nDistribución REAL de las combinaciones en la Muestra (M):")
sample_actual_distribution.toPandas()


--- Creando las Particiones (Mi) ---



Distribución REAL de las combinaciones en la Muestra (M):


verified_purchase vine   count  percentage
0                 N    N   82555   11.932862
1                 N    Y    3583    0.517903
2                 Y    N  605672   87.546489
3                 Y    Y      19    0.002746

En esta última fase del proceso se formaliza la creación de las particiones `𝑀𝑖`, definidas como subconjuntos disjuntos de la muestra `𝑀`, cuya segmentación responde a las combinaciones únicas de las variables categóricas seleccionadas. Dado que la muestra fue construida de manera estratificada en función de estas mismas variables, el agrupamiento por sus valores permite recuperar de forma directa cada una de las particiones sin necesidad de duplicar datos ni mantener estructuras adicionales.

Este diseño no solo reduce el costo computacional, sino que ofrece flexibilidad metodológica: cada partición puede ser invocada y procesada dinámicamente según los requerimientos del análisis posterior. Por ejemplo, acceder a la partición 
`𝑀𝑖` correspondiente a `verified_purchase='Y'` y `vine='N'` permite trabajar exclusivamente con ese segmento de usuarios, aislando su comportamiento o características para estudios comparativos.

Al verificar la distribución real de las combinaciones dentro de la muestra generada, se confirma que las proporciones han sido exitosamente conservadas. Las diferencias con respecto a los valores esperados son mínimas y atribuibles al redondeo y a la disponibilidad efectiva de registros dentro de ciertos estratos, sin comprometer la validez del enfoque.

Este paso final garantiza que cada partición sea estadísticamente coherente con su representación en la población `𝑃`, eliminando riesgos de sesgo por subrepresentación o sobreajuste a grupos específicos. Este diseño habilita un análisis posterior más sólido y confiable, ajustado a los principios fundamentales de muestreo en escenarios de datos masivos.

# 2) Construcción Train – Test
---

* **Instrucciones**: Construcción del conjunto de entrenamiento y prueba. Para este paso se asume que `𝑀` = {`𝑀𝑖`: `𝑀𝑖` es una partición derivada de las variables de caracterización de la población} generada en el paso anterior. Para construir el conjunto de entrenamiento y prueba, se debe de calcular el porcentaje de división a usar, de tal forma que al dividir cada `𝑀𝑖` en un conjunto de entrenamiento (**𝑇𝑟𝑖**) y prueba (**𝑇𝑠𝑖**), no se inyecten sesgos que desvíen la probabilidad de ocurrencia de los patrones en cada nueva partición. Para ello, deberás de retomar la estrategia de muestreo propuesta en el paso 4 de la Actividad 3 del módulo 4. Se debe de cuidar que `𝑇𝑟𝑖 Ç 𝑇𝑠𝑖 = Æ`, además de que la unión de todas las particiones es igual a `𝑀`.

In [ ]:
# Definir la proporción de división
train_ratio = 0.7
test_ratio = 1 - train_ratio
seed = 42

# Diccionarios para almacenar los conjuntos de entrenamiento y prueba por partición
train_partitions = {} # { (verified_purchase, vine): DataFrame_Tri }
test_partitions = {}  # { (verified_purchase, vine): DataFrame_Tsi }

# Obtener todas las combinaciones únicas de las variables categóricas en la muestra (M)
distinct_combinations = sample_df.select(categorical_cols).distinct().collect()
distinct_combinations

[Row(verified_purchase='N', vine='N'),
 Row(verified_purchase='N', vine='Y'),
 Row(verified_purchase='Y', vine='N'),
 Row(verified_purchase='Y', vine='Y')]

La definición de una proporción de **70/30** para los conjuntos de entrenamiento y prueba responde a una práctica comúnmente aceptada en el ámbito del análisis de datos, ya que permite un equilibrio adecuado entre aprendizaje y validación. El uso de una semilla asegura la reproducibilidad del proceso, lo cual es fundamental en entornos experimentales. 

Por otro lado, la identificación explícita de las combinaciones presentes en la muestra evita errores asociados al redondeo o a la ausencia de estratos poco frecuentes, garantizando que la división se realice únicamente sobre particiones efectivamente representadas. El resultado obtenido (cuatro combinaciones únicas) confirma que la muestra mantiene la diversidad estructural necesaria para una segmentación estratificada válida.

In [ ]:
# Iterar sobre cada combinación para crear las particiones Mi y dividirlas
for row in distinct_combinations:
    combo_values = tuple(row[col] for col in categorical_cols)
    print(f"\nProcesando partición Mi para: {combo_values}")

    # Aislamos la Partición Mi con apoyo de un filtro
    filter_condition = None
    for i, col_name in enumerate(categorical_cols):
        if filter_condition is None:
            filter_condition = (F.col(col_name) == combo_values[i])
        else:
            filter_condition = filter_condition & (F.col(col_name) == combo_values[i])

    mi_df = sample_df.filter(filter_condition)
    mi_count = mi_df.count()
    print(f"\tTamaño de Mi: {mi_count} registros")

    # Dividimos la Partición Mi en Tri y Tsi
    if mi_count > 0:
        # randomSplit distribuye los datos aleatoriamente según las proporciones dadas
        tri, tsi = mi_df.randomSplit([train_ratio, test_ratio], seed=seed)

        # Almacenar las particiones resultantes
        train_partitions[combo_values] = tri
        test_partitions[combo_values] = tsi

        print(f"\tTamaño de Tri: {tri.count()} registros")
        print(f"\tTamaño de Tsi: {tsi.count()} registros")
    else:
        print(f"\tPartición Mi vacía. No se generarán Tri y Tsi.")


Procesando partición Mi para: ('N', 'N')


	Tamaño de Mi: 82555 registros


	Tamaño de Tri: 57790 registros


	Tamaño de Tsi: 24765 registros

Procesando partición Mi para: ('N', 'Y')


	Tamaño de Mi: 3583 registros


	Tamaño de Tri: 2455 registros


	Tamaño de Tsi: 1128 registros

Procesando partición Mi para: ('Y', 'N')


	Tamaño de Mi: 605672 registros


	Tamaño de Tri: 424081 registros


	Tamaño de Tsi: 181591 registros

Procesando partición Mi para: ('Y', 'Y')


	Tamaño de Mi: 19 registros


	Tamaño de Tri: 16 registros


	Tamaño de Tsi: 3 registros


Con este código se hace la segmentación de la muestra representativa `𝑀` en subconjuntos de entrenamiento (`𝑇𝑟𝑖`) y prueba (`𝑇𝑠𝑖`) para cada una de las particiones `𝑀𝑖`, definidas previamente según las combinaciones únicas de las variables categóricas `verified_purchase` y `vine`. Esta estrategia garantiza que el proceso de entrenamiento y evaluación se lleve a cabo respetando la estructura interna de la población original, evitando así la introducción de sesgos derivados de una partición aleatoria no estratificada.

El método empleado consiste en filtrar individualmente cada estrato `𝑀𝑖`, y aplicar la función `randomSplit` de PySpark para generar las correspondientes divisiones con una proporción **70/30**. Esto permite mantener la representatividad estadística dentro de cada combinación de atributos. Los resultados muestran que la distribución se ha respetado incluso en estratos de baja frecuencia, como el caso de la combinación `('Y', 'Y')`, donde se preservó una división razonable de 16 observaciones para entrenamiento y 3 para prueba.

Esta partición controlada por estrato es especialmente valiosa en contextos donde existen desequilibrios severos entre categorías, ya que permite evaluar el desempeño del modelo de forma equitativa sobre todos los grupos relevantes. Además, al evitar el solapamiento entre `𝑇𝑟𝑖` y `𝑇𝑠𝑖`, se salvaguarda la independencia entre fases de entrenamiento y validación, requisito esencial para asegurar la robustez metodológica de los resultados.

In [19]:
# Verificamos las condiciones Tri Ç Tsi = Æ (no solapamiento) y la unión de todas las particiones = M
# Cabe destcar que la condición Tri Ç Tsi = Æ se garantiza por la función `randomSplit` de Spark
print("--- Verificación de las Condiciones ---")

# Unión de todos los Tri
all_tri_union = spark.createDataFrame([], sample_df.schema)
for combo, tri_df in train_partitions.items():
    all_tri_union = all_tri_union.unionAll(tri_df)

# Unión de todos los Tsi
all_tsi_union = spark.createDataFrame([], sample_df.schema)
for combo, tsi_df in test_partitions.items():
    all_tsi_union = all_tsi_union.unionAll(tsi_df)

# Unión de todos los Tri y Tsi
# Usamos distinct para manejar posibles duplicados si no hay garantía estricta
final_union_m = all_tri_union.unionAll(all_tsi_union).distinct()

print(f"Tamaño de la muestra original (M): {sample_df.count()}")
print(f"Tamaño de la unión de todos los Tri y Tsi: {final_union_m.count()}")

# Si los conteos coinciden, la condición se cumple.
if sample_df.count() == final_union_m.count():
    print("¡La condición de unión de particiones = M se cumple! Los conteos coinciden.")
else:
    print("Advertencia: Los conteos de la unión de particiones y la muestra M no coinciden.")

--- Verificación de las Condiciones ---


Tamaño de la muestra original (M): 691829


Tamaño de la unión de todos los Tri y Tsi: 691829


¡La condición de unión de particiones = M se cumple! Los conteos coinciden.


Una vez realizadas las divisiones `𝑇𝑟𝑖` y `𝑇𝑠𝑖` correspondientes a cada partición `𝑀𝑖`, se llevó a cabo un proceso de verificación para confirmar que se cumplan dos condiciones fundamentales de integridad en el diseño experimental: la no superposición entre conjuntos de entrenamiento y prueba, es decir, `𝑇𝑟𝑖 ∩ 𝑇𝑠𝑖 = ∅`, y la exhaustividad de la muestra, es decir, que la unión de todas las particiones `𝑇𝑟𝑖 ∪ 𝑇𝑠𝑖` sea equivalente al conjunto original `𝑀`.

La primera condición se garantiza implícitamente mediante el uso de la función `randomSplit` de PySpark, la cual está diseñada para generar subconjuntos mutuamente excluyentes a partir de un conjunto dado. En cuanto a la segunda condición, se construyó la unión de todas las particiones generadas y se aplicó una operación `distinct` para asegurar la unicidad de los registros. Al comparar el tamaño de esta unión con el total de registros de `𝑀`, se comprobó que ambos coinciden perfectamente, lo cual valida que no hubo pérdida ni duplicación de datos en el proceso de división.

# 3) Selección de métricas para medir calidad de resultados
---

Con el objetivo de evaluar la calidad de los resultados obtenidos a partir de los modelos implementados, es fundamental seleccionar cuidadosamente las **métricas de desempeño**, considerando no solo la naturaleza del algoritmo (supervisado o no supervisado), sino también el volumen de datos procesado y las implicaciones computacionales de cada métrica en un entorno de Big Data.

En el caso del modelo `RandomForestClassifier`, al tratarse de un enfoque supervisado orientado a la clasificación, se seleccionaron métricas estándar pero robustas que permiten evaluar el desempeño general y por clase del modelo. Específicamente, se consideran las siguientes: **accuracy, precision, recall y F1-score**, todas en su versión ponderada. Estas métricas permiten analizar no solo el número de aciertos globales, sino también la capacidad del modelo para distinguir correctamente entre las clases positivas y negativas, lo cual es especialmente relevante ante posibles desbalances en los datos. Adicionalmente, se utilizará la **matriz de confusión** como instrumento visual y cuantitativo que permite comprender la naturaleza de los errores cometidos (falsos positivos y falsos negativos), lo cual enriquece el análisis más allá de los valores promedio.

Por otro lado, para el modelo `KMeans`, al ser un algoritmo no supervisado de agrupamiento, las métricas tradicionales de clasificación no resultan aplicables. En su lugar, se optó por la utilización del **Silhouette Score**, el cual mide qué tan similares son los puntos dentro de un mismo clúster en comparación con puntos de otros clústers. Esta métrica, además de ser ampliamente aceptada en tareas de agrupamiento, resulta computacionalmente viable en PySpark y permite inferir si el número de clústers seleccionado es coherente con la estructura de los datos. Complementariamente, se evaluará la **coherencia semántica de los clústers obtenidos mediante inspección manual y análisis de centroides promedio**, especialmente al comparar los resultados con la variable de salida conocida, lo que permite explorar la alineación entre los patrones descubiertos de manera no supervisada y las etiquetas originales.

La selección de estas métricas responde a la doble necesidad de garantizar interpretabilidad y precisión en el análisis, al tiempo que se respeta la escalabilidad exigida por el tratamiento de grandes volúmenes de datos. La diversidad de enfoques implementados (supervisado y no supervisado) permite explorar diferentes dimensiones del comportamiento de los datos y establecer comparaciones que fortalecen la comprensión del problema y la calidad de las soluciones construidas.

# 4) Entrenamiento de Modelos de Aprendizaje
---

## 4.1) Preprocesamiento

In [20]:
from pyspark.ml.feature import StringIndexer, VectorAssembler, StandardScaler
from pyspark.ml import Pipeline

# Convertir las columnas categóricas a índices numéricos
indexer_verified = StringIndexer(inputCol="verified_purchase", outputCol="verified_purchase_index", handleInvalid="skip")
indexer_vine = StringIndexer(inputCol="vine", outputCol="vine_index", handleInvalid="skip")

label_col = "sentiment"
label_indexer = StringIndexer(inputCol=label_col, outputCol="label")

In [21]:
# Manejo de valores nulos 
initial_cols_to_check_for_nulls = [
    "helpful_votes", 
    "total_votes", 
    "verified_purchase", 
    "vine", 
    "sentiment" # Label
]

sample_df_cleaned = sample_df.na.drop(subset=initial_cols_to_check_for_nulls)
print(f"Registros después de eliminar nulos: {sample_df_cleaned.count()} (original: {sample_df.count()})")

Registros después de eliminar nulos: 691829 (original: 691829)


No se detectaron nulos, por lo que no se tiene que re-ejecutar el muestreo y la creación de particiones sobre sample_df_cleaned.

In [ ]:
# Definición de las columnas que serán ensambladas en el vector de características
feature_cols_for_assembler = [
    "helpful_votes", 
    "total_votes", 
    "verified_purchase_index", 
    "vine_index"
]

vector_assembler = VectorAssembler(inputCols=feature_cols_for_assembler, outputCol="raw_features")

In [ ]:
# Código para escalar las características
scaler = StandardScaler(
    inputCol="raw_features", 
    outputCol="scaled_features", 
    withStd=True, 
    withMean=False
)

## 4.2) Entrenamiento y evaluación de modelos

In [26]:
from pyspark.ml.classification import RandomForestClassifier
from pyspark.ml.clustering import KMeans
from pyspark.ml.evaluation import MulticlassClassificationEvaluator, ClusteringEvaluator

# Diccionarios para almacenar los resultados de la evaluación
rf_evaluation_results = {}
kmeans_evaluation_results = {}

# Definimos los modelos con hiperparámetros iniciales (se tunearán después)
rf_model_instance = RandomForestClassifier(
    featuresCol="scaled_features", 
    labelCol="label", 
    numTrees=20, 
    maxDepth=5, 
    seed=42
)

kmeans_model_instance = KMeans(
    featuresCol="scaled_features", 
    k=2, 
    seed=42
)

# Definimos los evaluadores
rf_evaluator = MulticlassClassificationEvaluator(
    labelCol="label", 
    predictionCol="prediction"
)

kmeans_evaluator = ClusteringEvaluator(
    featuresCol="scaled_features", 
    predictionCol="prediction", 
    metricName="silhouette"
)

In [28]:
# Bucle para entrenar y evaluar en cada partición
for combo_values, tri_df in train_partitions.items():
    tsi_df = test_partitions[combo_values]
    print(f"\nProcesando partición: {combo_values} (Train: {tri_df.count()}, Test: {tsi_df.count()})")

    # --- RandomForestClassifier ---
    print(f"\tEntrenando RandomForestClassifier para {combo_values}...")
    rf_pipeline = Pipeline(stages=[
        indexer_verified,
        indexer_vine,
        vector_assembler,
        scaler,
        label_indexer,
        rf_model_instance
    ])

    # Ejecutar pipeline de RF
    rf_fitted_pipeline = rf_pipeline.fit(tri_df)

    # Realizar predicciones en el conjunto de prueba
    rf_predictions = rf_fitted_pipeline.transform(tsi_df)

    # Evaluar RandomForestClassifier
    accuracy = rf_evaluator.evaluate(rf_predictions, {rf_evaluator.metricName: "accuracy"})
    f1_score = rf_evaluator.evaluate(rf_predictions, {rf_evaluator.metricName: "f1"})
    weighted_precision = rf_evaluator.evaluate(rf_predictions, {rf_evaluator.metricName: "weightedPrecision"})
    weighted_recall = rf_evaluator.evaluate(rf_predictions, {rf_evaluator.metricName: "weightedRecall"})

    print(f"\tRandomForest Metrics:")
    print(f"\t\tAccuracy: {accuracy:.4f}")
    print(f"\t\tF1-Score (Weighted): {f1_score:.4f}")
    print(f"\t\tPrecision (Weighted): {weighted_precision:.4f}")
    print(f"\t\tRecall (Weighted): {weighted_recall:.4f}")

    # Matriz de Confusión
    print("\n\tMatriz de Confusión (RandomForest):")
    confusion_pdf = rf_predictions.groupBy("label", "prediction").count().orderBy("label", "prediction").toPandas()
    display(confusion_pdf)

    rf_evaluation_results[combo_values] = {
        "accuracy": accuracy,
        "f1_score": f1_score,
        "weighted_precision": weighted_precision,
        "weighted_recall": weighted_recall,
        "confusion_matrix": confusion_pdf.to_dict('records') # Almacenar como lista de dicts
    }

    # --- KMeans ---
    print(f"\tEntrenando KMeans para {combo_values}...")
    kmeans_pipeline = Pipeline(stages=[
        indexer_verified,
        indexer_vine,
        vector_assembler,
        scaler,
        kmeans_model_instance
    ])

    # Ejecutar pipeline de KMeans
    kmeans_fitted_pipeline = kmeans_pipeline.fit(tri_df)

    # Realizar predicciones de clústeres en la misma partición de entrenamiento para evaluación
    # (KMeans es no supervisado, evaluamos la estructura de los clústeres en los datos que usó para aprender)
    kmeans_predictions = kmeans_fitted_pipeline.transform(tri_df)

    # Evaluar KMeans
    silhouette_score = kmeans_evaluator.evaluate(kmeans_predictions)
    # WCSS (Within-Cluster Sum of Squared Errors) se obtiene del modelo KMeans ajustado
    wcss = kmeans_fitted_pipeline.stages[-1].summary.trainingCost

    print(f"\tKMeans Metrics:")
    print(f"\t\tSilhouette Score: {silhouette_score:.4f}")
    print(f"\t\tWCSS: {wcss:.2f}")

    kmeans_evaluation_results[combo_values] = {
        "silhouette_score": silhouette_score,
        "wcss": wcss
    }

print("\n--- Entrenamiento y evaluación por partición completados. ---")
print("Resultados de RandomForest:", rf_evaluation_results)
print("Resultados de KMeans:", kmeans_evaluation_results)


Procesando partición: ('N', 'N') (Train: 57790, Test: 24765)
	Entrenando RandomForestClassifier para ('N', 'N')...


	RandomForest Metrics:
		Accuracy: 0.6609
		F1-Score (Weighted): 0.5271
		Precision (Weighted): 0.6790
		Recall (Weighted): 0.6609

	Matriz de Confusión (RandomForest):


label  prediction  count
0    0.0         0.0  16348
1    0.0         1.0      8
2    1.0         0.0   8389
3    1.0         1.0     20

	Entrenando KMeans para ('N', 'N')...


	KMeans Metrics:
		Silhouette Score: 0.9999
		WCSS: 50143.41



Procesando partición: ('N', 'Y') (Train: 2455, Test: 1128)
	Entrenando RandomForestClassifier para ('N', 'Y')...


	RandomForest Metrics:
		Accuracy: 0.7757
		F1-Score (Weighted): 0.6777
		Precision (Weighted): 0.6017
		Recall (Weighted): 0.7757

	Matriz de Confusión (RandomForest):


label  prediction  count
0    0.0         0.0    875
1    1.0         0.0    253

	Entrenando KMeans para ('N', 'Y')...


	KMeans Metrics:
		Silhouette Score: 0.9992
		WCSS: 1383.19



Procesando partición: ('Y', 'N') (Train: 424081, Test: 181591)
	Entrenando RandomForestClassifier para ('Y', 'N')...


	RandomForest Metrics:
		Accuracy: 0.7788
		F1-Score (Weighted): 0.6820
		Precision (Weighted): 0.6065
		Recall (Weighted): 0.7788

	Matriz de Confusión (RandomForest):


label  prediction   count
0    0.0         0.0  141424
1    1.0         0.0   40167

	Entrenando KMeans para ('Y', 'N')...


	KMeans Metrics:
		Silhouette Score: 1.0000
		WCSS: 103277.31



Procesando partición: ('Y', 'Y') (Train: 16, Test: 3)
	Entrenando RandomForestClassifier para ('Y', 'Y')...


	RandomForest Metrics:
		Accuracy: 1.0000
		F1-Score (Weighted): 1.0000
		Precision (Weighted): 1.0000
		Recall (Weighted): 1.0000

	Matriz de Confusión (RandomForest):


label  prediction  count
0    0.0         0.0      3

	Entrenando KMeans para ('Y', 'Y')...


	KMeans Metrics:
		Silhouette Score: 0.9059
		WCSS: 5.15

--- Entrenamiento y evaluación por partición completados. ---
Resultados de RandomForest: {('N', 'N'): {'accuracy': 0.6609327680193822, 'f1_score': 0.5271011415626842, 'weighted_precision': 0.6790089713851148, 'weighted_recall': 0.6609327680193823, 'confusion_matrix': [{'label': 0.0, 'prediction': 0.0, 'count': 16348}, {'label': 0.0, 'prediction': 1.0, 'count': 8}, {'label': 1.0, 'prediction': 0.0, 'count': 8389}, {'label': 1.0, 'prediction': 1.0, 'count': 20}]}, ('N', 'Y'): {'accuracy': 0.775709219858156, 'f1_score': 0.6777289739150141, 'weighted_precision': 0.601724793772949, 'weighted_recall': 0.775709219858156, 'confusion_matrix': [{'label': 0.0, 'prediction': 0.0, 'count': 875}, {'label': 1.0, 'prediction': 0.0, 'count': 253}]}, ('Y', 'N'): {'accuracy': 0.7788051169936836, 'f1_score': 0.6819604963590836, 'weighted_precision': 0.6065374102555452, 'weighted_recall': 0.7788051169936836, 'confusion_matrix': [{'label': 0.0, 'pre

## 4.3) Ajuste de hiperparámetros

In [ ]:
from pyspark.ml.tuning import ParamGridBuilder, CrossValidator

# Seleccionamos una partición representativa para el ajuste de hiperparámetros:
# Para este ejemplo, tomaremos la primera partición de entrenamiento disponible.
if not train_partitions:
    print("No hay particiones de entrenamiento disponibles para el ajuste de hiperparámetros.")
else:
    first_combo_key = list(train_partitions.keys())[0]
    tuning_data_df = train_partitions[first_combo_key]
    print(f"Realizando ajuste de hiperparámetros en la partición: {first_combo_key} (Tamaño: {tuning_data_df.count()})")

    print("\n  Ajustando Hiperparámetros para RandomForestClassifier...")

    # Instanciamos un nuevo RandomForestClassifier (sin numTrees y maxDepth específicos aquí, se definirán en ParamGrid)
    rf_tune_model_instance = RandomForestClassifier(
        featuresCol="scaled_features", 
        labelCol="label", 
        seed=42
    )

    # Reconstruimos el pipeline de RF para tuning, usando la instancia `rf_tune_model_instance`
    rf_tune_pipeline = Pipeline(stages=[
        indexer_verified,
        indexer_vine,
        vector_assembler,
        scaler,
        label_indexer,
        rf_tune_model_instance
    ])

    # Definir la cuadrícula de parámetros a probar
    rf_paramGrid = (ParamGridBuilder()
                    .addGrid(rf_tune_model_instance.numTrees, [10, 50]) 
                    .addGrid(rf_tune_model_instance.maxDepth, [5, 10])  
                    .addGrid(rf_tune_model_instance.minInstancesPerNode, [1, 5]) 
                    .build())

    # Crear el CrossValidator
    rf_crossval = CrossValidator(estimator=rf_tune_pipeline,
                                 estimatorParamMaps=rf_paramGrid,
                                 evaluator=MulticlassClassificationEvaluator(labelCol="label", metricName="f1"), # Usar F1 para seleccionar el mejor modelo
                                 numFolds=3, # Número de folds para validación cruzada
                                 seed=42,
                                 parallelism=spark.sparkContext.defaultParallelism) # Usar todos los cores disponibles

    # Entrenar el CrossValidator (puede tardar un tiempo considerable)
    print("    Iniciando Cross-Validation para RandomForest... (Esto puede tomar tiempo)")
    rf_cv_model = rf_crossval.fit(tuning_data_df) # Entrenar en los datos de tuning
    best_rf_pipeline_model = rf_cv_model.bestModel # El mejor pipeline entrenado

    print("    Mejores Hiperparámetros encontrados para RandomForestClassifier:")
    best_rf_params = best_rf_pipeline_model.stages[-1].extractParamMap()
    print(f"      numTrees: {best_rf_params.get(rf_tune_model_instance.numTrees)}")
    print(f"      maxDepth: {best_rf_params.get(rf_tune_model_instance.maxDepth)}")
    print(f"      minInstancesPerNode: {best_rf_params.get(rf_tune_model_instance.minInstancesPerNode)}")

    # Puedes almacenar este `best_rf_pipeline_model` para usarlo en futuras inferencias o para re-evaluar las particiones.


    # --- Ajuste de Hiperparámetros para KMeans ---
    print("\n  Ajustando Hiperparámetros para KMeans (k)...")

    # Instanciar un nuevo KMeans (sin k específico aquí)
    kmeans_tune_model_instance = KMeans(featuresCol="scaled_features", seed=42)

    # Reconstruir el pipeline de KMeans para tuning
    kmeans_tune_pipeline = Pipeline(stages=[
        indexer_verified,
        indexer_vine,
        vector_assembler,
        scaler,
        kmeans_tune_model_instance
    ])

    # Definir la cuadrícula de parámetros para k
    kmeans_paramGrid = (ParamGridBuilder()
                        .addGrid(kmeans_tune_model_instance.k, [2, 3, 4, 5]) # Probar diferentes números de clústeres
                        .build())

    # Crear el CrossValidator
    kmeans_crossval = CrossValidator(estimator=kmeans_tune_pipeline,
                                     estimatorParamMaps=kmeans_paramGrid,
                                     evaluator=ClusteringEvaluator(featuresCol="scaled_features", metricName="silhouette"), # Usar Silhouette para seleccionar el mejor k
                                     numFolds=3, # Número de folds
                                     seed=42,
                                     parallelism=spark.sparkContext.defaultParallelism)

    # Entrenar el CrossValidator
    print("    Iniciando Cross-Validation para KMeans... (Esto puede tomar tiempo)")
    kmeans_cv_model = kmeans_crossval.fit(tuning_data_df) # Entrenar en los datos de tuning
    best_kmeans_pipeline_model = kmeans_cv_model.bestModel # El mejor pipeline entrenado

    print("    Mejor número de clústeres (k) encontrado para KMeans:")
    best_kmeans_k = best_kmeans_pipeline_model.stages[-1].getK()
    print(f"      k: {best_kmeans_k}")

    # Puedes almacenar este `best_kmeans_pipeline_model` para usarlo en futuras inferencias o análisis de clústeres.

## 4.4) Almacenamiento y carga de modelos

In [ ]:
base_model_path = "/ml_models/"

# Guardamos el mejor modelo RandomForest
if 'best_rf_pipeline_model' in locals():
    rf_model_save_path = path.join(base_model_path, "random_forest_sentiment_classifier")
    best_rf_pipeline_model.write().overwrite().save(rf_model_save_path)
    print("Modelo RandomForest guardado exitosamente.")

# Guardamos el mejor modelo KMeans
if 'best_kmeans_pipeline_model' in locals():
    kmeans_model_save_path = path.join(base_model_path, "kmeans_review_clusters")
    best_kmeans_pipeline_model.write().overwrite().save(kmeans_model_save_path)
    print("Modelo KMeans guardado exitosamente.")

In [ ]:
from pyspark.ml import PipelineModel

# Ejemplo de cómo cargar el modelo (RandomForestClassifier)
loaded_rf_model = PipelineModel.load(rf_model_save_path)
print("Modelo RandomForest cargado exitosamente")

# Ejemplo de cómo cargar el modelo (KMeans)
loaded_kmeans_model = PipelineModel.load(kmeans_model_save_path)
print("Modelo KMeans cargado exitosamente")

# 5) Análisis de resultados
---